File: src/core/action.ts
import type { MatchTarget } from '../types'
import type { TmuxClient } from '../infra/tmux'

type CursorTarget = Pick<MatchTarget, 'line' | 'col' | 'primary'>

export const createMoveCursorCommands = (
  paneId: string,
  target: CursorTarget,
): string[][] => {
  const commands: string[][] = [['send-keys', '-X', '-t', paneId, 'top-line']]

  if (target.line > 1) {
    commands.push([
      'send-keys',
      '-X',
      '-N',
      String(target.line - 1),
      '-t',
      paneId,
      'cursor-down',
    ])
  }

  commands.push(['send-keys', '-X', '-t', paneId, 'start-of-line'])

  const right = target.col + target.primary
  if (right > 0) {
    commands.push([
      'send-keys',
      '-X',
      '-N',
      String(right),
      '-t',
      paneId,
      'cursor-right',
    ])
  }

  return commands
}

export const moveCopyCursor = async (
  tmux: TmuxClient,
  paneId: string,
  target: CursorTarget,
): Promise<void> => {
  for (const command of createMoveCursorCommands(paneId, target)) {
    await tmux.run(command)
  }
}

File: src/action.rs
use anyhow::Result;

use crate::tmux::{s, tmux};
use crate::types::MatchTarget;

pub fn move_copy_cursor(pane_id: &str, target: &MatchTarget) -> Result<()> {
    tmux(&s(&["send-keys", "-X", "-t", pane_id, "top-line"]))?;
    if target.candidate.line > 1 {
        tmux(&s(&[
            "send-keys",
            "-X",
            "-N",
            &target.candidate.line.saturating_sub(1).to_string(),
            "-t",
            pane_id,
            "cursor-down",
        ]))?;
    }
    tmux(&s(&["send-keys", "-X", "-t", pane_id, "start-of-line"]))?;
    let right = target.candidate.char_col + target.primary_char;
    if right > 0 {
        tmux(&s(&[
            "send-keys",
            "-X",
            "-N",
            &right.to_string(),
            "-t",
            pane_id,
            "cursor-right",
        ]))?;
    }
    Ok(())
}
